src/script/interpreter.cpp

```diff
@@ -1258,65 +1349,207 @@ class CTransactionSignatureSerializer
+}
+
+/** Compute the (single) SHA256 of the concatenation of all scriptPubKeys spent by a tx. */
+uint256 GetSpentScriptsHash(const std::vector<CTxOut>& outputs_spent)
+{
+    CHashWriter ss(SER_GETHASH, 0);
+    for (const auto& txout : outputs_spent) {
+        ss << txout.scriptPubKey;
+    }
+    return ss.GetSHA256();
+}
+
+
 } // namespace
 
 template <class T>
-void PrecomputedTransactionData::Init(const T& txTo)
+void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut> spent_outputs)
 {
-    assert(!m_ready);
+    assert(!m_spent_outputs_ready);
 
-    // Cache is calculated only for transactions with witness
-    if (txTo.HasWitness()) {
-        hashPrevouts = GetPrevoutHash(txTo);
-        hashSequence = GetSequenceHash(txTo);
-        hashOutputs = GetOutputsHash(txTo);
+    m_spent_outputs = std::move(spent_outputs);
+    if (!m_spent_outputs.empty()) {
+        assert(m_spent_outputs.size() == txTo.vin.size());
+        m_spent_outputs_ready = true;
+    }
+
+    // Determine which precomputation-impacting features this transaction uses.
+    bool uses_bip143 = false;
+    bool uses_bip341 = false;
+    for (size_t inpos = 0; inpos < txTo.vin.size(); ++inpos) {
+        if (!txTo.vin[inpos].scriptWitness.IsNull()) {
+            if (m_spent_outputs_ready && m_spent_outputs[inpos].scriptPubKey.size() == 2 + WITNESS_V1_TAPROOT_SIZE &&
+                m_spent_outputs[inpos].scriptPubKey[0] == OP_1) {
+                // Treat every native witness v1 spend as a Taproot spend. This only works if spent_outputs was
+                // provided as well, but if it wasn't, actual validation will fail anyway.
+                uses_bip341 = true;
+            } else {
+                // Treat every spend that's not known to native witness v1 as a Witness v0 spend. This branch may
+                // also be taken for unknown witness versions, but it is harmless, and being precise would require
+                // P2SH evaluation to find the redeemScript.
+                uses_bip143 = true;
+            }
+        }
     }
 
-    m_ready = true;
+    if (uses_bip143 || uses_bip341) {
+        // Computations shared between both sighash schemes.
+        m_bip341_prevouts_hash = GetPrevoutHash(txTo);
+        m_bip341_sequences_hash = GetSequenceHash(txTo);
+        m_bip341_outputs_hash = GetOutputsHash(txTo);
+    }
+    if (uses_bip143) {
+        hashPrevouts = SHA256Uint256(m_bip341_prevouts_hash);
+        hashSequence = SHA256Uint256(m_bip341_sequences_hash);
+        hashOutputs = SHA256Uint256(m_bip341_outputs_hash);
+        m_bip143_ready = true;
+    }
+    if (uses_bip341) {
+        m_bip341_spent_amounts_hash = GetSpentAmountsHash(m_spent_outputs);
+        m_bip341_spent_scripts_hash = GetSpentScriptsHash(m_spent_outputs);
+        m_bip341_ready = true;
+    }
 }
 
 template <class T>
 PrecomputedTransactionData::PrecomputedTransactionData(const T& txTo)
 {
-    Init(txTo);
+    Init(txTo, {});
```

Comment:
I would very much like that, but I think we inevitably need a few states in it at least.

The `PrecomputedTransactionData` objects used for block validation are constructed once up front, and then initialized simultaneously with other checks in `CheckInputScripts`. This means they need some sort of "uninitialized" state which can detect incorrect use before initialization. A cleaner solution would be to move the construction of these objects out of `CheckInputScripts` and do it up front, so they can be constructed with all available information once and for all. I think that makes sense, but it's a bigger change to validation logic that I'd like to avoid in this PR.

The other reason is indeed libconsensus, which has no API for taproot validation. There is also no problem, as it doesn't expose a taproot validation flag (yet). A future extension to it would add a new call that lets you pass in the spent UTXOs, and let you enable taproot validation. But as of right now, changing things to require the UTXOs whenever initializing the `PrecomputedTransactionData` would mean that libconsensus would need to construct mock UTXOs and pass those in... which I think is worse than having an explicit uninitialized state that'd trigger an error on misuse.